core/client/src/main/java/alluxio/client/block/UnderStoreBlockInStream.java

```diff
@@ -28,78 +29,67 @@
+    /**
+     * @return an input stream to under storage
+     * @throws IOException if an IO exception occurs
+     */
+    InputStream create() throws IOException;
+
+    /**
+     * Closes the factory, releasing any resources it was holding.
+     *
+     * @throws IOException if an IO exception occurs
+     */
+    void close() throws IOException;
+  }
+
+  /**
+   * Creates an under store block in stream which will read from the streams created by the given
+   * {@link UnderStoreStreamFactory}. The stream will be set to the beginning of the block.
    * @param initPos the initial position
    * @param length the length of this current block (allowed to be {@link Constants#UNKNOWN_SIZE})
    * @param fileBlockSize the block size for the file
-   * @param ufsPath the under file system path
+   * @param underStoreStreamFactory a factory for getting input streams from the under storage
+   * @throws IOException if an IO exception occurs while creating the under storage input stream
    */
-  protected UnderStoreBlockInStream(long initPos, long length, long fileBlockSize, String ufsPath) {
+  public UnderStoreBlockInStream(long initPos, long length, long fileBlockSize,
+      UnderStoreStreamFactory underStoreStreamFactory) throws IOException {
     mInitPos = initPos;
     mLength = length;
     mFileBlockSize = fileBlockSize;
-    mUfsPath = ufsPath;
-  }
-
-  /**
-   * Factory for creating {@link UnderStoreBlockInStream}s.
-   */
-  @ThreadSafe
-  public static final class Factory {
-
-    private Factory() {} // prevent instantiation
-
-    /**
-     * Creates an under store block in stream, if ufs operation delegation is enabled, the stream
-     * will read from an Alluxio worker, if not the stream will be directly from an under storage
-     * system client. The stream will be set to the beginning of the block.
-     *
-     * @param blockStart the start position of the block stream relative to the entire file
-     * @param length length of this block
-     * @param blockSize the block size of the file
-     * @param path the path of the file in the under storage
-     * @return a stream which can access data from blockStart to blockStart + length
-     * @throws IOException if an error occurs creating the stream
-     */
-    public static UnderStoreBlockInStream create(long blockStart, long length, long blockSize,
-        String path) throws IOException {
-      UnderStoreBlockInStream stream;
-      if (Configuration.getBoolean(Constants.USER_UFS_DELEGATION_ENABLED)) {
-        stream = new DelegatedUnderStoreBlockInStream(blockStart, length, blockSize, path);
-      } else {
-        stream = new DirectUnderStoreBlockInStream(blockStart, length, blockSize, path);
-      }
-      stream.setUnderStoreStream(0);
-      return stream;
-    }
+    mUnderStoreStreamFactory = underStoreStreamFactory;
+    setUnderStoreStream(0);
   }
 
   @Override
   public void close() throws IOException {
+    mUnderStoreStreamFactory.close();
```

Comment:
Doesn't need to be addressed in this PR, but one possible optimization is closing the under store stream factory at the file level instead of at the block level for the ufs streams.